src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java

```diff
@@ -29,168 +29,415 @@
+import jenkins.scm.impl.trait.Selection;
+import jenkins.scm.impl.trait.WildcardSCMHeadFilterTrait;
 import org.apache.commons.lang.StringUtils;
 import org.jenkins.ui.icon.Icon;
 import org.jenkins.ui.icon.IconSet;
 import org.jenkinsci.Symbol;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
 import org.kohsuke.stapler.AncestorInPath;
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.kohsuke.stapler.DataBoundSetter;
 import org.kohsuke.stapler.QueryParameter;
 
 public class BitbucketSCMNavigator extends SCMNavigator {
 
-    private final String repoOwner;
-    private String credentialsId;
-    private String checkoutCredentialsId;
-    private String pattern = ".*";
-    private boolean autoRegisterHooks = false;
-    private String bitbucketServerUrl;
-    /**
-     * Ant match expression that indicates what branches to include in the retrieve process.
-     */
-    private String includes = "*";
+    private static final Logger LOGGER = Logger.getLogger(BitbucketSCMSource.class.getName());
 
-    /**
-     * Ant match expression that indicates what branches to exclude in the retrieve process.
-     */
-    private String excludes = "";
+    @NonNull
+    private String serverUrl;
+    @CheckForNull
+    private String credentialsId;
+    @NonNull
+    private final String repoOwner;
+    @NonNull
+    private List<SCMTrait<? extends SCMTrait<?>>> traits;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    private transient String checkoutCredentialsId;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    private transient String pattern;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    private transient boolean autoRegisterHooks;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    private transient String includes;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    private transient String excludes;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    private transient String bitbucketServerUrl;
 
 
     @DataBoundConstructor
     public BitbucketSCMNavigator(String repoOwner) {
+        this.serverUrl = BitbucketCloudEndpoint.SERVER_URL;
         this.repoOwner = repoOwner;
+        this.traits = new ArrayList<>();
         this.credentialsId = null; // highlighting the default is anonymous unless you configure explicitly
-        this.checkoutCredentialsId = BitbucketSCMSource.DescriptorImpl.SAME;
+        this.traits.add(new BranchDiscoveryTrait(true, false));
+        this.traits.add(new OriginPullRequestDiscoveryTrait(EnumSet.of(ChangeRequestCheckoutStrategy.MERGE)));
+        this.traits.add(new ForkPullRequestDiscoveryTrait(EnumSet.of(ChangeRequestCheckoutStrategy.MERGE),
```

Comment:
Well we should be able to make it work with some secondary changes in the Mercurial plugin... just wanting to keep those changes out of this 12k LoC PR ;-)